Apps/Src/UpdateVelocity.c

```diff
@@ -1,67 +1,117 @@
-                car_state->DesiredMotorCurrent = 1.0f;
+    MotorController_Init(1.0f);
+
+    static uint8_t prevBrakePedalPercent;
+
+    //Prev refers to it being pressed before
+    static State cruzEnablePrev = OFF;
+    static State cruzSetPrev = OFF;    
+
+    //State refers to whether we consider the enable and set to be on or off
+    static State cruzEnableState = OFF;
+    static State cruzSetState = OFF;
+    
+    float desiredVelocity = 0;
+    float desiredMotorCurrent = 0;
+    while (1)
+    {
+        uint8_t accelPedalPercent = Pedals_Read(ACCELERATOR);
+        uint8_t brakePedalPercent = Pedals_Read(BRAKE);
+
+        // If the brake is changed, then the brake lights might need to be changed
+        if (brakePedalPercent != prevBrakePedalPercent)
+        {
+            if(brakePedalPercent >= UNTOUCH_PEDALS_PERCENT){
+                Lights_Set(BrakeLight, ON);
+            }
+            else{
+                Lights_Set(BrakeLight, OFF);
             }
-        } else { // Not in regen or cruise control mode
-            car_state->CRSet = ACCEL;
-            car_state->CruiseControlEnable = OFF;
         }
 
-        if(car_state->CRSet == ACCEL) {
-            car_state->DesiredVelocity = MAX_VELOCITY;
-            car_state->DesiredMotorCurrent = convertPedaltoMotorPercent(car_state->AccelPedalPercent);
+        prevBrakePedalPercent = brakePedalPercent;
+
+        State regenPressed = Switches_Read(REGEN_SW);
+
+        //Current refers to it currently being pressed
+        State cruzEnableCurr = Switches_Read(CRUZ_EN);
+        State cruzSetCurr = Switches_Read(CRUZ_ST);
+
+        //The cruzEnableState is toggled on the rising edge of the button press
+        if(cruzEnableCurr && !cruzEnablePrev){
+            cruzEnablePrev = ON;
+            cruzEnableState = (State) ((int)cruzEnableState ^ 1);
+        }else if(!cruzEnableCurr){
+            cruzEnablePrev = OFF;
         }
 
+        //The cruzSetState is toggled on the rising edge of the button press
+        if(cruzSetCurr && !cruzSetPrev){
+            cruzSetPrev = ON;
+            cruzSetState = (State) ((int)cruzSetState ^ 1);
+        }else if(!cruzSetCurr){
+            cruzSetPrev = OFF;
+        }
+
+        //Regen will be level sensitive meaning the user will regen brake by holding down a button
+        if(regenPressed && RegenEnable && (brakePedalPercent < UNTOUCH_PEDALS_PERCENT)){
+            desiredVelocity = 0;
+            desiredMotorCurrent = REGEN_CURRENT;
+        } else if(cruzEnableState && cruzSetState && brakePedalPercent < UNTOUCH_PEDALS_PERCENT 
+            && accelPedalPercent < UNTOUCH_PEDALS_PERCENT){
+            desiredVelocity = MotorController_ReadVelocity();
+            desiredMotorCurrent = 1.0f;
+        }
+        else if(brakePedalPercent >= UNTOUCH_PEDALS_PERCENT){
+            desiredVelocity = 0;
+            desiredMotorCurrent = 0;
+        }
+        else{
+            if(!Switches_Read(REV_SW)){
```

Comment:
```suggestion
            if(Switches_Read(REV_SW)){
```
If the switch is in reverse then we should go backwards